vnext/Microsoft.ReactNative/Utils/TextHitTestUtils.cpp

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "TextHitTestUtils.h"
+
+#include <UI.Xaml.Documents.h>
+
+namespace Microsoft::ReactNative {
+
+struct BidirectionalTextBoundary {
+  xaml::Documents::TextPointer ltrPointer{nullptr};
+  xaml::Documents::TextPointer rtlPointer{nullptr};
+  winrt::Rect ltrRect;
+  winrt::Rect rtlRect;
+};
+
+/**
+ * Finds the character closest to X = 0 for the first line. I.e., the
+ * leftmost character when flow direction is left-to-right.
+ * ┌─────> X ↓
+ * │ (x,y)  ┌─<─┐<┌─>─┐
+ * │ ┌──────┘   └─┘   │
+ * │ │                │
+ * ↓ │     ┌──────────┘
+ * Y └─────┘
+ */
+static BidirectionalTextBoundary FindStartRTLBoundary(
+    const xaml::Documents::TextPointer &start,
+    const xaml::Documents::TextPointer &end,
+    const winrt::Rect &target) {
+  auto textPointer = start;
+  auto L = start.Offset();
+  auto R = end.Offset();
+  while (L <= R) {
+    const auto m = /* floor */ (L + R) / 2;
+    const auto relativeOffset = m - textPointer.Offset();
```

Comment:
I may have mentioned this, but not all glyphs will be the same amount of `wchar_t`s, e.g. in CJK (Chinese, Japanese, Korean), some characters take up two UTF-16 characters. Same with a lot of the newer emojis. I *believe* TextPointer might operate on UTF16 units and not characters, so the logic might be slightly off for texts that contain complex emoji sequences, for example. Doesn't sound like a big deal for most usage, but if FBM is impacted by usage of complex emoji, it might be worth verifying.

Reply:
There actually doesn't seem to be a width value on the TextPointer object, so I suspect this won't be a problem. We calculate width by incrementing the pointer and finding the left most position of the next character.

This works pretty well in most cases, but there are some notable edge cases (e.g., the last character in line-wrapped text is not always hit testable because of this).